### ISSUE TITLE
Edge feathering for flow for a fix direction

### ISSUE THREAD

**Issue by nitroman123** [crest-ocean/crest#851]
Please describe the desired feature/behavior.

This will help simulating dynamic waves when the ship is is at low speeds. This will help a lot for waterjets in the near future I want to add and the dynamic waves to be simulated like real when the ship has low speeds . This will look very realistic 

Supporting images/mockups, or example reference videos are very helpful to communicate ideas.

https://user-images.githubusercontent.com/62849221/121402064-22a2ba80-c962-11eb-8d7c-81cec68306e6.mp4

**Comment by daleeidd** [comment]
Thanks.

One workaround until we look into adding this is making use of the texture input is an option. Here is a radial gradient which won't have a seam.

![Radial](https://user-images.githubusercontent.com/5249806/121408522-5b687280-c915-11eb-8b6f-ce602494034e.png)

**Comment by nitroman123** [comment]
the result is amazing !!! really

**Comment by nitroman123** [comment]
@daleeidd @huwb  After some experiment with programming some custom properties I found that the flow float2 attribute they x and y values must have a different value for every single rotation in order to simulate the correct value (pushing the waves backwards or regardless the rotation this could help a lot with my case and will give a unique touch for the game)   

![Screenshot_381](https://user-images.githubusercontent.com/62849221/121798177-8280b600-cc2d-11eb-9e78-5e0af8e839c9.png)

**Comment by huwb** [comment]
Thanks for this, looks interesting. we can help but i think i need to see the issue in motion to understand your latest posts. or if you can share the effect with us we can work on it with you.

**Comment by nitroman123** [comment]
@huwb yes here i made a 3 min video to see , hope this helps , i tweak the values runtime to see the correction it makes , its a very nice effect and for me and i think everyone love to have it cause it simulates the real sea behavior especially for ships. The combination of fixed direction and the 2 texture could be awesome .

https://youtu.be/VGu_Crjg9V8

**Comment by nitroman123** [comment]
Any updates for this i really love to have it to make some nice videos... this feature will give a very nice feeling on the water pushing ?

**Comment by daleeidd** [comment]
@nitroman123 Unfortunately, no. All time has gone towards the next release as it was a big update.

**Comment by nitroman123** [comment]
@huwb  @daleeidd  Ive made some experiments and i found the values for the degrees based on the rotation , i was trying to make a script to control the X and Z values. I made the X and Z float variables and i found the right values for the X and Z .

Basically the values goes from to for each rotation 

degrees 0 x = 0  , z = 1 
degrees 90 x = 1  , z = 0 
degrees 180 x = 0  , z = -1 
degrees 270 x = -1  , z = 0
degrees 360 x = 0  , z = 1 

tell me if i cant make something the help the procedure cause i really want it !!! 

Thanks in advance

**Comment by nitroman123** [comment]
So after 1 day of trying i made a script that controls the values for the shader to be rotated with the right numbers 

was so confusing 

`
using UnityEngine;
using UnityEngine.UI;
public class Flow : MonoBehaviour
{
    public Material flowmat;
    [SerializeField] float X;
    [SerializeField] float Z = 1;
    public Transform players;
    public float to45;
    public float to90;
    public float to135;
    public float to180;
    // Update is called once per frame
    void Update()
    {
        //Application.targetFrameRate = 70;
        float rot = players.eulerAngles.y;
        if (rot >= 0 && rot <= 45)
        {
            X = Mathf.Clamp01(players.rotation.normalized.y * to45);
            Z = Mathf.Clamp01(1 - players.rotation.normalized.y * to45);
        }
        if (rot > 45 && rot <= 90)
        {
            X = Mathf.Clamp01(players.rotation.normalized.y * to90);
            Z = Mathf.Clamp01(1 - players.rotation.normalized.y * to90);
        }
        if (rot > 90 && rot <= 135)
        {
            X = 1 - Mathf.Clamp01(players.rotation.normalized.y * to135);
            Z = -Mathf.Clamp01(players.rotation.normalized.y * to135);
        }
        if (rot > 135 && rot <= 180)
        {
            X = 1 - Mathf.Clamp01(players.rotation.normalized.y * to180);
            Z = -Mathf.Clamp01(players.rotation.normalized.y * to180);
        }
        flowmat.SetFloat("_FlipX", X);
        flowmat.SetFloat("_FlipZ", Z);
    }
}
`
assign the material  and the modified shader 

`// Crest Ocean System

// This file is subject to the MIT License as seen in the root of this folder structure (LICENSE)

Shader "Crest/Inputs/Flow/Add Flow Map"
{
	Properties
	{
		_FlowMap("Flow Map", 2D) = "white" {}
		_Strength( "Strength", float ) = 1
		_FlipX("Flip X", Float) = 0
		_FlipZ("Flip Z", Float) = 0
	}

	SubShader
	{
		Tags { "Queue" = "Transparent" }
		Blend One One

		Pass
		{
			CGPROGRAM
			#pragma vertex Vert
			#pragma fragment Frag

			#pragma shader_feature_local _FLIPX_ON
			#pragma shader_feature_local _FLIPZ_ON

			#include "UnityCG.cginc"

			sampler2D _FlowMap;

			CBUFFER_START(CrestPerOceanInput)
			float4 _FlowMap_ST;
			float _Strength;
			float3 _DisplacementAtInputPosition;
			float _FlipZ;
			float _FlipX;
			CBUFFER_END

			struct Attributes
			{
				float3 positionOS : POSITION;
				float2 uv : TEXCOORD0;
			};

			struct Varyings
			{
				float4 positionCS : SV_POSITION;
				float2 uv : TEXCOORD0;
			};

			Varyings Vert(Attributes input)
			{
				Varyings o;

				float3 worldPos = mul(unity_ObjectToWorld, float4(input.positionOS, 1.0)).xyz;
				// Correct for displacement
				worldPos.xz -= _DisplacementAtInputPosition.xz;
				o.positionCS = mul(UNITY_MATRIX_VP, float4(worldPos, 1.0));
				
				o.uv = TRANSFORM_TEX(input.uv, _FlowMap);
				return o;
			}

			float4 Frag(Varyings input) : SV_Target
			{
				float2 flow = tex2D(_FlowMap, input.uv).xy - 0.5;

				flow.x *= _FlipX;


				flow.y *= _FlipZ;


				return float4(flow * _Strength, 0.0, 0.0);
			}

			ENDCG
		}
	}
}
`

**Comment by nitroman123** [comment]
So after 1 day of trying i made a script that controls the values for the shader to be rotated with the right numbers 

script : 

https://pastebin.com/JpfB4FPa

assign the material  and the modified shader 

https://pastebin.com/2kNcg5us

**Comment by daleeidd** [comment]
Nice work. Have a video of it in action? Very interested in seeing it.

**Comment by nitroman123** [comment]
it doesnt work correctly for all rotations and imagine that if you want to make a flow input look at the side of the ship (Bow thruster) the rotations change dramatically i will make a sort video and send it over in a bit

**Comment by nitroman123** [comment]
here is the video [https://youtu.be/vru4gtWMCTc] it is almost getting the right values but i would very costly to make for a lot of points and more points mean more script etc... I really want this feature it will give a very big impact on the visuals and a unique feature for a ship sim game

**Comment by daleeidd** [comment]
The video is private and I cannot view it.

**Comment by nitroman123** [comment]
sry for this now you should be able to see it [https://youtu.be/vru4gtWMCTc](url)

**Comment by huwb** [comment]
Still can't seem to access it. Has it worked for you dale?

**Comment by daleeidd** [comment]
> Still can't seem to access it. Has it worked for you dale?

It does but I had to copy and paste the link manually. I've updated both links and they will work when clicked now.

**Comment by daleeidd** [comment]
> sry for this now you should be able to see it https://youtu.be/vru4gtWMCTc

Thanks. Looks great. Interesting approach to get that effect.

**Comment by nitroman123** [comment]
Is very nice effect to show that forces push the water backwards etc. This is a must effect in my ship sim game. Its getting close to release next year and having this effect will be a miracle. If a had a fixed direction with the smoothness of the add flow map is all i need. 

Sry for the link don't know what was wrong with that.

**Comment by daleeidd** [comment]
Test branch here: [test/feature/flow-input-improvements](https://github.com/wave-harmonic/crest/tree/test/feature/flow-input-improvements). Spoke with @nitroman123 on Discord with feedback. The rotation works with a quad as I just took the code from the spline. I'll need to work out how to make it user friendly before it can be merged. Will do a PR soon for the UV feathering as that works well.

**Comment by daleeidd** [comment]
Feathering has been added for 4.15. Rotation has been moved to its own branch: https://github.com/wave-harmonic/crest/tree/feature/flow-rotation
